src/typemap/src/fs_type_conversion.rs

```diff
@@ -0,0 +1,173 @@
+/// argument is absolute.
+///
+/// ## Arguments:
+/// origp: path to normalize.
+/// cageid: cage ID of the `origp`
+///
+/// ## Returns:
+/// A `PathBuf` representing the normalized absolute path.
+pub fn normpath(origp: PathBuf, cageid: u64) -> PathBuf {
+    let cage = cage::get_cage(cageid).unwrap();
+    //If path is relative, prefix it with the current working directory, otherwise populate it with rootdir
+    let mut newp = if origp.is_relative() {
+        (**cage.cwd.read()).clone()
+    } else {
+        PathBuf::from("/")
+    };
+
+    for comp in origp.components() {
+        match comp {
+            //if we have a normal path component, push it on to our normed path
+            Component::Normal(_) => {
+                newp.push(comp);
+            }
+
+            //if we have a .. path component, pop the last component off our normed path
+            Component::ParentDir => {
+                newp.pop();
+            }
+
+            //if we have a . path component (Or a root dir or a prefix(?)) do nothing
+            _ => {}
+        };
+    }
+    newp
+}
+
+/// This function first normalizes the path, then add `LIND_ROOT` at the beginning.
+/// This function is mostly used by path argument translation function in `syscall_conv`
+///
+/// ## Arguments:
+///     - cageid: used for normalizing path
+///     - path: the user seen path
+///
+/// ## Returns:
+///     - c_path: path location from host's perspective
+pub fn add_lind_root(cageid: u64, path: &str) -> CString {
+    // Convert data type from &str into *const i8
+    let relpath = normpath(convpath(path), cageid);
+    let relative_path = relpath.to_str().unwrap();
+
+    let full_path = format!("{}{}", LIND_ROOT, relative_path);
+    let c_path = CString::new(full_path).unwrap();
+    c_path
+}
+
+/// Translate a received virtual file descriptor (`virtual_fd`) to real kernel file descriptor.
+///
+/// This function is used to translate a per-cage virtual file descriptor into the actual
+/// kernel-level file descriptor managed by the `fdtables`.
+///
+/// Optionally, when the `secure` feature is enabled, this function will verify that the
+/// `arg_cageid` (the cage that owns the fd being translated) is allowed to interact with the
+/// caller's `cageid` (the requesting cage).
+///
+/// ## Arguments:
+/// virtual_fd: The virutal file descriptor
+/// arg_cageid: The cage that owns the virtual fd
+/// cageid: The cage ID of current caller (only used when `secure` mode is enabled)
+///
+/// ## Returns:
+/// underlying kernel file descriptor
+pub fn convert_fd_to_host(virtual_fd: u64, arg_cageid: u64, cageid: u64) -> i32 {
+    #[cfg(feature = "secure")]
+    {
+        if !validate_cageid(arg_cageid, cageid) {
+            return -1;
+        }
+    }
+    // Find corresponding virtual fd instance from `fdtable` subsystem
+    let wrappedvfd = fdtables::translate_virtual_fd(arg_cageid, virtual_fd);
+    if wrappedvfd.is_err() {
+        return -9;
```

Comment:
this and the -1 above seem like magic numbers, what do they signify? best way is macros but if not then comment